### ISSUE TITLE
native image on windows : dynamically load sunmscapi

### ISSUE THREAD

**Issue by mttcr** [oracle/graal#1410]
Hello,
I'm trying to build a windows executable that makes use of sunmscapi. the native image command line is : "native-image myProgram --enable-all-security-services --no-fallback"
I read JCA-SECURITY-SERVICES.md about the Native implementations paragraph.
But i can't build the program and dynamically link it later, I have the following error:
```
  Creating library c:\Users\aaa\Downloads\graalvm-ce-windows-amd64-19.0.2\test\myprogram.lib and object c:\Users\aaa\Downloads\graalvm-ce-windows-amd64-19.0.2\test\myprogram.exp
myprogram.obj : error LNK2001: unresolved external symbol Java_sun_security_ec_ECDSASignature_signDigest
myprogram.obj : error LNK2001: unresolved external symbol Java_sun_security_ec_ECDSASignature_verifySignedDigest
myprogram.obj : error LNK2001: unresolved external symbol Java_sun_security_mscapi_KeyStore_loadKeysOrCertificateChains
myprogram.obj : error LNK2001: unresolved external symbol Java_sun_security_mscapi_PRNG_generateSeed
myprogram.obj : error LNK2001: unresolved external symbol Java_sun_security_mscapi_RSASignature_generatePublicKeyBlob
myprogram.obj : error LNK2001: unresolved external symbol Java_sun_security_mscapi_RSASignature_importPublicKey
myprogram.obj : error LNK2001: unresolved external symbol Java_sun_security_smartcardio_PCSC_SCardEstablishContext
c:\Users\aaa\Downloads\graalvm-ce-windows-amd64-19.0.2\test\myprogram.exe : fatal error LNK1120: 7 unresolved externals

        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593)
        at java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1005)
        at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:458)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:292)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:430)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:112)
Caused by: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 2
```

What could be the solution?

**Comment by lilili87222** [comment]
I figured out a way. it can resolve this temporary .
Edit graalvm-ce-19.0.2\jre\lib\security\java.security

security.provider.1=sun.security.provider.Sun
security.provider.2=sun.security.rsa.SunRsaSign
security.provider.3=org.jcp.xml.dsig.internal.dom.XMLDSigRI
# security.provider.3=sun.security.ec.SunEC
security.provider.4=com.sun.net.ssl.internal.ssl.Provider
security.provider.5=com.sun.crypto.provider.SunJCE
security.provider.6=sun.security.jgss.SunProvider
security.provider.7=com.sun.security.sasl.Provider
#security.provider.8=org.jcp.xml.dsig.internal.dom.XMLDSigRI
# security.provider.9=sun.security.smartcardio.SunPCSC
# security.provider.10=sun.security.mscapi.SunMSCAPI

**Comment by ebourg** [comment]
I confirm this works, thank you for the workaround.

**Comment by lilili87222** [comment]
It may caused by not load .lib file 
I add 

`getNativeLibraries().addLibrary("sunec", true);`

it stop ec problem ,but mscapi.lib file not found .

tested windows 7 amd64, graalvm 19.2.1

**Comment by mttcr** [comment]
Hello,
I gave a try today to launch a native image generated binary that uses windows sunmsc api with `graalvm-ce-java11-20.2.0-dev`
The executable generation works, the command line is : 
`native-image --verbose --static -H:-CheckToolchain --enable-all-security-services --enable-https --enable-http...`
When I run the executable, I have the following error:
```
19:49:43.162 - java.lang.UnsatisfiedLinkError: sun.security.mscapi.CKeyStore.loadKeysOrCertificateChains(Ljava/lang/String;)V [symbol: Java_sun_security_mscapi_CKeyStore_loadKeysOrCertificateChains or Java_sun_security_mscapi_CKeyStore_loadKeysOrCertificateChains__Ljava_lang_String_2]
        at com.oracle.svm.jni.access.JNINativeLinkage.getOrFindEntryPoint(JNINativeLinkage.java:153)
        at com.oracle.svm.jni.JNIGeneratedMethodSupport.nativeCallAddress(JNIGeneratedMethodSupport.java:57)
        at sun.security.mscapi.CKeyStore.loadKeysOrCertificateChains(CKeyStore.java)
        at sun.security.mscapi.CKeyStore.engineLoad(CKeyStore.java:711)
        at sun.security.mscapi.CKeyStore$MY.engineLoad(CKeyStore.java:57)
        at java.security.KeyStore.load(KeyStore.java:1479)
```
while calling `KeyStore.getInstance("Windows-MY", "SunMSCAPI").load(null, null)`

FYI, I did launch the java program with `%GRAALVMPATH%\bin\java -agentlib:native-image-agent=config-merge-dir....` in order to have config files.

Any idea on how to solve this?

**Comment by pejovica** [comment]
This was fixed a while ago in #3300.
